rts/lra/lra-test/lra-test-basic/src/test/java/io/narayana/lra/arquillian/FailedLRAIT.java

```diff
@@ -0,0 +1,232 @@
+            Assert.assertTrue(response.hasEntity());
+
+            return URI.create(response.readEntity(String.class));
+        } finally {
+            if (response != null) {
+                response.close();
+            }
+        }
+    }
+
+    /**
+     * Validate whether or not a given LRA is in a particular state.
+     * Also validate that the corresponding record is removed.
+     *
+     * @param lra the LRA whose state is to be validated
+     * @param state the state that the target LRA should be in
+     * @return true if the LRA is in the target state and that its log was successfully removed
+     * @throws Exception if the state cannot be determiined
+     */
+    private boolean validateStateAndRemove(URI lra, LRAStatus state) throws Exception {
+        List<JSONObject> failedRecords = getFailedRecords(lra);
+        for (JSONObject failedLRA : failedRecords) {
+            String lraId = (String) failedLRA.get("lraId");
+            lraId = lraId.replaceAll("\\\\", "");
+            if (lraId.contains(lra.toASCIIString())) {
+                String status = (String) failedLRA.get("status");
+                if (status.equals(state.name())) {
+                    // remove the failed LRA
+                    Assert.assertEquals("Could not remove log", Response.Status.OK.getStatusCode(), removeFailedLRA(lra));
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    // look up LRAs that are in a failed state (ie FailedToCancel or FailedToClose)
+    private List<JSONObject> getFailedRecords(URI lra) throws Exception {
+        Response response = null;
+        String recoveryUrl = String.format("http://%s:%d/%s",
+                lra.getHost(), lra.getPort(), LRAConstants.RECOVERY_COORDINATOR_PATH_NAME);
+
+        try {
+            response = client.target(new URI(recoveryUrl)).path("failed").request().get();
+
+            Assert.assertTrue("Missing response body when querying for failed LRAs", response.hasEntity());
+            String failedLRAs = response.readEntity(String.class);
+
+            JSONArray jsonArray = new JSONArray(failedLRAs);
+            List<JSONObject> failedList = new ArrayList<>();
+
+            for (int i = 0; i < jsonArray.length(); i++) {
+                failedList.add(new JSONObject(jsonArray.getString(i)));
+            }
+
+            return failedList;
+        } finally {
+            if (response != null) {
+                response.close();
+            }
+        }
+    }
+
+    // ask the recovery coordinator to delete its log for an LRA
+    private int removeFailedLRA(URI lra) throws URISyntaxException, UnsupportedEncodingException {
+        Response response = null;
+        String recoveryUrl = String.format("http://%s:%d/%s",
+                lra.getHost(), lra.getPort(), LRAConstants.RECOVERY_COORDINATOR_PATH_NAME);
+        String txId = URLEncoder.encode(lra.toASCIIString(), "UTF-8");
```

Comment:
I would expect to also see the operation getFailedLRAs and removeFailed(URI) in NarayanaLRAClient. WDYT?

Reply:
Currently we only use NarayanaLRAClient internally for the participant JAX-RS filter and some of our tests.
The API to the coordinator is the OpenAPI doc.

So there would be no consumers of the two methods that you are proposing. But I do agree that these two methods would be useful in a client API at the spec level.

Reply:
I would expect narayana users to use the client directly. If they won't be spec only of course. Should we take this in another JIRA or will you update this PR?

Reply:
We don't require a new JIRA. Clients will use the REST api.

Reply:
Okay I will add it

Reply:
We don't have any methods in NarayanaLRAClient that return collections of LRAs so let's take that to another JIRA if you don't mind.

Reply:
ah sure because we didn't change it since the spec version. However, I think this is a good idea to add all operations exposed by coordinator to the programmatic API also. Later when we again introduce client API in the spec we will just have more methods in narayana client so users can chose to inject spec version or narayana one. Please create a new JBTM, we should add also other methods that the coordinator exposes as JAX-RS endpoints. This can be resolved.

Reply:
https://issues.redhat.com/browse/JBTM-3303